centreon/src/Core/Common/Infrastructure/Repository/ApiCallIterator.php

```diff
@@ -0,0 +1,132 @@
+
+use Psr\Log\LoggerInterface;
+use Symfony\Contracts\HttpClient\Exception\ClientExceptionInterface;
+use Symfony\Contracts\HttpClient\Exception\RedirectionExceptionInterface;
+use Symfony\Contracts\HttpClient\Exception\ServerExceptionInterface;
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+/**
+ * @template T
+ *
+ * @implements \IteratorAggregate<int, T>&\Countable
+ */
+class ApiCallIterator implements \IteratorAggregate, \Countable
+{
+    use ApiResponseTrait;
+
+    private int $nbrElements = 0;
+
+    /** @var list<T> */
+    private array $entitiesCache = [];
+
+    private bool $isCachedCreated = false;
+
+    /**
+     * @param HttpClientInterface $httpClient
+     * @param string $url
+     * @param array<string, mixed> $options
+     * @param int $maxItemsByRequest
+     * @param \Closure $entityFactory
+     * @param LoggerInterface $logger
+     */
+    public function __construct(
+        private readonly HttpClientInterface $httpClient,
+        private readonly string $url,
+        private readonly array $options,
+        private readonly int $maxItemsByRequest,
+        private readonly \Closure $entityFactory,
+        private readonly LoggerInterface $logger,
+    ) {
+    }
+
+    /**
+     * @throws TransportExceptionInterface
+     * @throws ServerExceptionInterface
+     * @throws RedirectionExceptionInterface
+     * @throws ClientExceptionInterface
+     */
+    public function getIterator(): \Traversable
+    {
+        $this->createCache();
+        foreach ($this->entitiesCache as $entity) {
+            yield $entity;
+        }
+    }
+
+    /**
+     * @throws TransportExceptionInterface
+     * @throws ServerExceptionInterface
+     * @throws RedirectionExceptionInterface
+     * @throws ClientExceptionInterface
+     */
+    public function count(): int
+    {
+        $this->createCache();
+
+        return $this->nbrElements;
+    }
+
+    /**
+     * @throws TransportExceptionInterface
+     * @throws ServerExceptionInterface
+     * @throws RedirectionExceptionInterface
+     * @throws ClientExceptionInterface
+     */
+    private function createCache(): void
+    {
+        if (! $this->isCachedCreated) {
+            $fromPage = 1;
+            $totalPage = 0;
+            do {
+                $url = $this->url . sprintf('?limit=%d&page=%d', $this->maxItemsByRequest, $fromPage);
+                $this->logger->debug('Call API', ['url' => $this->url]);
+                /** @var array{meta: array{total: int}, result: array<string, mixed>} $response */
+                $response = $this->getResponseOrFail($this->httpClient->request('GET', $url, $this->options));
+                if ($this->nbrElements === 0) {
+                    $this->nbrElements = (int) $response['meta']['total'];
+                    $totalPage = (int) ceil($this->nbrElements / $this->maxItemsByRequest);
```

Comment:
I don't any check about `maxItemsByRequest` being not zero `0`.
Even the phpdoc about the property say `int` but might need a `positive-int` to be consistent.